src/datastore/postgres-store.ts

```diff
@@ -4270,6 +4330,341 @@ export class PgDataStore
     return events;
   }
 
+  async getFilteredAddressEvents(args: {
+    principal: string;
+    eventTypeFilter: DbEventTypeId[];
+    limit: number;
+    offset: number;
+  }) {
+    return this.queryTx(async client => {
+      const eventIndexStart = args.offset;
+      const eventIndexEnd = args.offset + args.limit - 1;
+      const eventsList: DbEvent[] = [];
+      for (const eventType of args.eventTypeFilter) {
+        switch (eventType) {
+          case DbEventTypeId.StxLock:
+            const stxLockResults = await client.query<StxLockEventQueryResult>(
+              `
+              SELECT
+                event_index, tx_id, tx_index, block_height, canonical, locked_amount, unlock_height, locked_address
+              FROM stx_lock_events
+              WHERE locked_address = $1 AND microblock_canonical = true AND event_index BETWEEN $2 AND $3
+              `,
```

Comment:
A few comments on these queries:
1. Querying for events in separate `query()` calls is a bad idea for pagination because you want to paginate the results, not filter against the `event_index` property of every table. You should switch to a single query in which you make `UNION` of all the results and then paginate over those **sorted** results (it will be very important to sort by `block_height, microblock_sequence, tx_index, event_index`).
2. Make sure to filter both `canonical` AND `microblock_canonical`

Reply:
I followed the same approach as this https://github.com/hirosystems/stacks-blockchain-api/blob/f1abc291d955525983d40da11f8f5c638cee756c/src/datastore/postgres-store.ts#L4077

should we look into this as well?

Reply:
Interesting @asimm241.

Where is that function being used? We could evaluate whether or not to refactor that one (later, in another PR) based on the current use.

Reply:
@rafaelcr I don't think we can do union here since all events have different columns.